src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java

```diff
@@ -1,40 +1,115 @@
 package ru.yandex.practicum.filmorate.controller;
 
-
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.http.ResponseEntity;
+import ru.yandex.practicum.filmorate.exception.NoParameterException;
 import ru.yandex.practicum.filmorate.model.Film;
 import org.springframework.web.bind.annotation.*;
-import ru.yandex.practicum.filmorate.storage.film.InMemoryFilmManager;
+import ru.yandex.practicum.filmorate.service.FilmService;
 
 import javax.validation.Valid;
+import javax.validation.constraints.Positive;
 
 import java.util.Collection;
 
-import static ru.yandex.practicum.filmorate.storage.Managers.getDefaultFilmManager;
-
-
 @RestController
+@RequestMapping("/films")
 public class FilmController {
+    FilmService filmService;
 
-    private final InMemoryFilmManager filmManager = getDefaultFilmManager();
+    @Autowired
+    public FilmController(FilmService filmService) {
+        this.filmService = filmService;
+    }
 
-    @PostMapping("/films")
+    @PostMapping
     public ResponseEntity<Film> addFilm(@Valid @RequestBody Film film) {
-        return filmManager.addFilm(film);
+        return filmService.addFilm(film);
     }
 
-    @PutMapping("/films")
+    @PutMapping
     public ResponseEntity<Film> updateFilm(@Valid @RequestBody Film film) {
-        return filmManager.updateFilm(film);
+        return filmService.updateFilm(film);
     }
 
-    @GetMapping("/films")
-    public Collection<Film> getAllFilm() {
-        return filmManager.getAllFilm();
+    @GetMapping("/{filmId}")
+    public ResponseEntity<Film> getFilmById(@PathVariable(required = false) Integer filmId) {
+        if (filmId == null) {
+            throw new NoParameterException("filmId");
+        }
+        if (filmId <= 0) {
+            throw new NoParameterException("filmId");
+        }
+        return filmService.getFilmById(filmId);
     }
 
-    @DeleteMapping("/films")
+    @GetMapping
+    public ResponseEntity<Collection<Film>> getAllFilm() {
+        return filmService.getAllFilm();
+    }
+
+    @DeleteMapping
     public ResponseEntity<String> removeAllFilm() {
-        return filmManager.removeAllFilm();
+        return filmService.removeAllFilm();
+    }
+
+    @DeleteMapping("/{filmId}")
+    public ResponseEntity<Film> removeFilmById(@PathVariable(required = false) Integer filmId) {
+        if (filmId == null) {
+            throw new NoParameterException("filmId");
+        }
+        if (filmId <= 0) {
+            throw new NoParameterException("filmId");
+        }
+        return filmService.removeFilmById(filmId);
+    }
+
+    @PutMapping("/{filmId}/like/{userId}")
+    public ResponseEntity<Film> addUserLikeByFilmId(@PathVariable(required = false) Integer filmId,
+                                                    @PathVariable(required = false) Integer userId) {
+        if (filmId == null) {
```

Comment:
Предлагаю отказаться от required=false, оставить по умолчанию аргументы как обязательные(ведь они правда нам нужны) а валидацию провести так же как и в dto. Смотри, мы можем просто так же повесить аннотацию, как и в наших user/film классах. Вот пример:
@PathVariable("name") @NotBlank @Size(max = 10) String username. Так же и тут, Positive или какая там аннотация нужна, уже не помню :) Нужно будет сделать так во всех местах с такой валидацией

Reply:
В следствии была удалена и замена на проверку if

Reply:
Можешь подсказать как сделать проверку сразу нескольких значений

mockMvc.perform(put("/users/1/friends/3"))
                .andExpect(status()
                        .isOk())
                .andExpect(jsonPath("$.friendsIds")
                        .value("[2,3]"));
                        
Что только не пробовал и лист.офф и Matchers и просто строка( в интернете так ничего не нашел( как это сделать так и не понял

Reply:
> Странно, но валидация POSITIVE вообще не работает( в тестах проходит id -1 до storage

Скорее всего, проблема самого фреймворка. Помню когда-то решал проблему с тем, что проверки не проходят при тестировании, точнее, игнорируется валидация. При этом в реальной работе все ок. Даже как-то решал, но вспомнить бы как...

Reply:
Попробуй просто убрать Optional, оставить Integer, и убрать Valie

Reply:
А по множествам проверок вот так:
        mockMvc.perform(delete("/users"))
                .andExpect(status()
                        .isResetContent())
                .andExpect(jsonPath("$.friendsIds")
                                .value("[2,3]"))


Скорее всего, ты просто не там ставил точку и у тебя просто не на том объекте из-за этого

Reply:
Или ты про массив? Массив чуть мудреннее, можешь воспользоваться вот этим ответом: https://stackoverflow.com/questions/25614593/asserting-array-of-arrays-with-jsonpath-and-spring-mvc

Reply:
А с валидацией в mockMvc - попробуй вот такую анноташку @WebMvcTest над классом тестов

Reply:
> > Странно, но валидация POSITIVE вообще не работает( в тестах проходит id -1 до storage
> 
> Скорее всего, проблема самого фреймворка. Помню когда-то решал проблему с тем, что проверки не проходят при тестировании, точнее, игнорируется валидация. При этом в реальной работе все ок. Даже как-то решал, но вспомнить бы как...

Да я уже только чего не пробовал,) и с @Valid, и с @Validated, и без него. Контроллер с фильмами без Опционал), тоже всё пробовал интернет копал ничего не помогло, надоело искать решение время на неё убил уйма, написал if и вуаля всё ловит. А эти Позитиве пусть лесом идут) В общем жалко что не работает готовая валидация(

Reply:
> А по множествам проверок вот так: mockMvc.perform(delete("/users")) .andExpect(status() .isResetContent()) .andExpect(jsonPath("$.friendsIds") .value("[2,3]"))
> 
> Скорее всего, ты просто не там ставил точку и у тебя просто не на том объекте из-за этого

Стоп какую такую точку?)
Так я так и делал с самого начала на интуиции, а оказалось не всё так гладко

mockMvc.perform(put("/users/1/friends/3"))
.andExpect(status()
.isOk())
.andExpect(jsonPath("$.friendsIds")
.value("[2,3]"));

Здесь все точки на месте, получаю такой получаю такой ответ от идеи, то что мл это не множество)

Got a list of values [2,3] instead of the expected single value [2,3]

Reply:
> Или ты про массив? Массив чуть мудреннее, можешь воспользоваться вот этим ответом: https://stackoverflow.com/questions/25614593/asserting-array-of-arrays-with-jsonpath-and-spring-mvc

Да не я про мой конкретный случай у там Set<Integer>

Reply:
> А с валидацией в mockMvc - попробуй вот такую анноташку @WebMvcTest над классом тестов

А что она будет проверять?

Reply:
> Или ты про массив? Массив чуть мудреннее, можешь воспользоваться вот этим ответом: https://stackoverflow.com/questions/25614593/asserting-array-of-arrays-with-jsonpath-and-spring-mvc

Да не я про мой конкретный случай у там Set

Я там был уже всё замылил уже)

Reply:
> А с валидацией в mockMvc - попробуй вот такую анноташку @WebMvcTest над классом тестов

А что она будет делать?

Reply:
> Или ты про массив? Массив чуть мудреннее, можешь воспользоваться вот этим ответом: https://stackoverflow.com/questions/25614593/asserting-array-of-arrays-with-jsonpath-and-spring-mvc

Да не я про мой конкретный случай у там Set

Я был там уже и не раз и не только там) много где копал) всё замылил по этому вопросу), спрашиваю не просто так же)

Reply:
> А с валидацией в mockMvc - попробуй вот такую анноташку @WebMvcTest над классом тестов

А что она будет делать?

Reply:
С ней все тесты падают test ignored , скажи почему так происходит?

Reply:
> С ней все тесты падают test ignored , скажи почему так происходит?

А фиг его знает честно говоря, сейчас смотрю тоже падает) Вообще падает потому что нельзя совмещать ее с другой аннотацией, которая в себе содержит аннотацию BootStrapWith. Поэтому теперь мне даже не вспомнить как заставить валидацию работать в тестах. Гугл подскажет в любом случае)

Reply:
> Да, через маппер я знаю, что можно) хотелось без использования оного научиться) На много меньше запись будет, в тестах и так много нужно написать)

Запись же будет одинакового размера, но гораздо более читаемая?